File: 2_Periodo/Algoritmos_e_Estruturas_de_Dados_2/TPs/TP01/TP01Q01/Palindromo.java
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q01 - 19 / 02 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : javac Palindromo.java
 *       Windows: javac Palindromo.java
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : java Palindromo
 *       Windows: java Palindromo
 *   
*/

class Palindromo
{
    /**
     *  Funcao para verificar se a entrada e' igual a "FIM" 
     *  @param s - String
     *  @return true se fim, false caso contrario.
     */
    public static boolean isFim( String s )
    {
        boolean result = false;
        if( s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M' )
        {
            result = true;
        } // end if
        return ( result );
    } // end isFim ( )

    /**
     *  Funcao para verificar se uma string e' um palindromo.
     *  @return true se palindromo, false caso contrario.
     *  @param s - String
    */
    public static boolean isPalindromo ( String s )
    {
        boolean result = true;
        int x = 0, y = 0;
        
        for ( x = 0, y = s.length( )-1; x < s.length( )/2; x=x+1, y=y-1 )
        {
            if ( s.charAt(x) != s.charAt(y) ) // verificar se o primeiro caractere e' diferente do ultimo
            {
                result = false;  
                x = s.length( ); // interromper a repeticao
            } // end if
        } // end for
        return ( result );
    } // end isPalindromo ( )

    /**
     *  Funcao principal.
     *  @param args
     */
    public static void main( String[] args ) 
    {
        MyIO.setCharset("UTF-8");
        String entrada = "";

        do 
        {
            entrada = MyIO.readLine();
            if( !isFim( entrada ) )
            {
                if( isPalindromo( entrada ) )
                {
                    MyIO.println( "SIM" );
                }
                else
                {
                    MyIO.println( "NAO" );
                } // end if
            } // end if
        } while ( isFim( entrada ) == false ); // end do while
    } // end main ( )
} // end class

File: 2_Periodo/Algoritmos_e_Estruturas_de_Dados_2/TPs/TP01/TP01Q02/Palindromo.c
/*
 *  -------------------------- Documentacao
 *  Pontificia Universidade Catolica de Minas Gerais
 *  Curso de Ciencia da Computacao
 *  Algoritmos e Estruturas de Dados II
 *   
 *  TP01Q02 - 22 / 02 / 2024
 *  Author: Vinicius Miranda de Araujo
 *   
 *  Para compilar em terminal (janela de comandos):
 *       Linux : gcc -o Palindromo Palindromo.c
 *       Windows: gcc -o Palindromo Palindromo.c
 *   
 *  Para executar em terminal (janela de comandos):
 *       Linux : ./Palindromo
 *       Windows: Palindromo
 *   
*/

// ---------------------------------------- Dependencias

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <wchar.h>   // wide-character types
#include <locale.h>  // location specific settings

// ---------------------------------------- Metodos

/**
 *  Funcao para verificar se a entrada e' igual a "FIM" 
 *  @param s - char*.
 *  @return true se fim, false caso contrario.
 */
bool isFim( char* s )
{
    bool result = false;
    if( strcmp( s, "FIM" ) == 0 )
    {
        result = true;
    }
    return ( result );
} // end isFim ( )

/**
 *  Funcao para verificar se uma string e' um palindromo.
 *  @return true se palindromo, false caso contrario.
 *  @param s - char*.
*/
bool isPalindromo ( char *s )
{
    bool result = true;
    int x = 0, y = 0;
    unsigned int length = strlen( s );

    for ( x = 0, y = length-1; x < length/2; x=x+1, y=y-1 )
    {
        if ( s[x] != s[y] ) // verificar se o primeiro caractere e' diferente do ultimo
        {
            result = false;  
            x = length; // interromper a repeticao
        } // end if
    } // end for
    return ( result );
} // end isPalindromo ( )

/**
 *  Funcao para ler linha da entrada padrao.
 *  @return string.
*/
char* readLine( )
{
    char *string = (char*) calloc ( 1000+1, sizeof(char) );
    if( string != NULL )
    {
        fgets( string, 1000+1, stdin );
        unsigned int length = strlen( string );
        // if( length > 0 && ( string[length-1] == '\n' || string[length-1] == '\r') )
        if( length > 0 && string[length-1] == '\n' )
        {
            string[length-1] = '\0';
        } // end if
    } // end if
    return ( string ); 
} // end readLine ( )

/**
 *  Funcao Principal
 *  @param agrc
 *  @param agrv
*/
int main( int argc, char const *argv[] )
{
    setlocale(LC_CTYPE, "UTF-8"); // setCharset
    char* entrada = NULL;

    do
    {
        entrada = readLine( );

        if( !isFim( entrada ) )
        {
            if( isPalindromo( entrada ) )
            {
                printf( "SIM\n" );
            }
            else
            {
                printf( "NAO\n" );
            } // end if
            free( entrada );
        } // end if
    } while ( !isFim( entrada ) );
    
    return ( 0 );
} // end main ( )
